keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+status on API server configuration objects has been tricky to design in the
+past, because of the following:
+
+- multiple active kube-apiservers (sometimes at identical versions, sometimes
+  skewed by one version during upgrade)
+- multiple active non-kube-apiserver servers (aggregated servers)
+- Note: 'NonStructural' relies on the fact that the status change is a function
+  of the resource then generation can be used
+  (https://github.com/kubernetes/kubernetes/commit/2cfc3c69dc7c17b2711af0168f39ed7f515675c2).
+  Can this approach be somehow extended?S
+
+Note that a policy configuration CRD may be referenced by the `spec.config` of
+multiple `ValidatingAdmissionPolicy` resources. Each of which may apply
+different policy validation rules using the same configuration.  For example,
+one `ValidatingAdmissionPolicy` might validate the containers declared in `pods`
+while another might validate the containers declared in the `podTemplate` of a
+`replicaSet`. As long as both are validating that the same policy configuration
+is being enforced, just in different ways, it is reasonable for both to share
+the same policy configuration resources.
+
+Alternative considered: Embed the configuration CRD schema directly into
+`ValidatingAdmissionPolicy`.
+
+Pros:
+
+- Policy author doesn't need to define the CRD. It can instead be generated for them.
+
+Cons:
+
+- This implies that the policy owns the configuration. But we have use cases
+  where multiple policies share the same configuration. For this model it seems
+  misleading and potentially to have both policies defining the configuration.
+
+##### Match Criteria
+
+During admission, the Kubernetes API server will validate the resource being
+admitted against all policy configuration resources that match the resource.
+
+Match criteria must be declared in the `spec.match` field of policy
+configuration resources (see `ReplicaLimit` in the above example).
+
+In order for policy configuration resources to declare match criteria, the
+corresponding configuration CRD schema must has a `spec.match` property. This
+property must conform to the below "matching schema template". This ensures that
+the match criteria is in the format that API server expects (the API server will
+be using duck typing here since there is no established way to do polymorphism
+across CRDs). The schemas of these fields in the configuration CRD may omit any
+optional properties; policy definition authors should only include the parts of
+the "match schema template" that are useful for configuring a particular policy.
+
+(Also, by allowing the "matching schema template" in configuration CRDs to be a
+omit optional properties, this API is future proofed against the addition of
+other match related properties in the future).
+
+"matching schema template":
+
+```go
+// TODO: Add this as a struct into the Kubernetes codebase so it can easily be
+// imported?
+type Match struct {
```

Comment:
Other common matching criteria include
1. which user or group is requesting
2. does the user or group have an authorizer permission like "use" for PSP
3. an indicator on the object being admitted that indicates which configuration to use.

Can we get a table of matching mechanisms and whether we'll support them or not?

Reply:
Yes, added.